src/stan/services/optimize/bfgs.hpp

```diff
@@ -0,0 +1,202 @@
+       *   the gradient
+       * @param[in] tol_param convergence tolerance on changes in parameter value
+       * @param[in] num_iterations maximum number of iterations
+       * @param[in] save_iterations indicates whether all the interations should
+       *   be saved to the parameter_writer
+       * @param[in] refresh how often to write output to message_writer
+       * @param[out] interrupt interrupt callback to be called every iteration
+       * @param[out] message_writer output for messages
+       * @param[out] init_writer Writer callback for unconstrained inits
+       * @param[out] parameter_writer output for parameter values
+       * @return stan::services::error_codes::OK if successful
+       */
+      template <class Model>
+      int bfgs(Model& model,
+               stan::io::var_context& init,
+               unsigned int random_seed,
+               unsigned int chain,
+               double init_radius,
+               double init_alpha,
+               double tol_obj,
+               double tol_rel_obj,
+               double tol_grad,
+               double tol_rel_grad,
+               double tol_param,
+               int num_iterations,
+               bool save_iterations,
+               int refresh,
+               callbacks::interrupt& interrupt,
+               callbacks::writer& message_writer,
+               callbacks::writer& init_writer,
+               callbacks::writer& parameter_writer) {
+        boost::ecuyer1988 rng = stan::services::util::rng(random_seed, chain);
+
+        std::vector<int> disc_vector;
+        std::vector<double> cont_vector
+          = stan::services::util::initialize(model, init, rng, init_radius,
+                                             false,
+                                             message_writer, init_writer);
+
+        std::stringstream bfgs_ss;
+        typedef stan::optimization::BFGSLineSearch
+          <Model, stan::optimization::BFGSUpdate_HInv<> > Optimizer;
+        Optimizer bfgs(model, cont_vector, disc_vector, &bfgs_ss);
+        bfgs._ls_opts.alpha0 = init_alpha;
+        bfgs._conv_opts.tolAbsF = tol_obj;
+        bfgs._conv_opts.tolRelF = tol_rel_obj;
+        bfgs._conv_opts.tolAbsGrad = tol_grad;
+        bfgs._conv_opts.tolRelGrad = tol_rel_grad;
+        bfgs._conv_opts.tolAbsX = tol_param;
+        bfgs._conv_opts.maxIts = num_iterations;
+
+        double lp = bfgs.logp();
+
+        std::stringstream msg;
+        msg << "Initial log joint probability = " << lp;
+        message_writer(msg.str());
+
+        std::vector<std::string> names;
+        names.push_back("lp__");
+        model.constrained_param_names(names, true, true);
+        parameter_writer(names);
+
+        if (save_iterations) {
+          std::vector<double> values;
+          msg.str("");
+          model.write_array(rng, cont_vector, disc_vector, values,
+                            true, true, &msg);
+          if (msg.str().length() > 0)
```

Comment:
Yes. We'll tackle the output issue after this is in. There are hacks in RStan now that parse output: I don't know how much it tries to do through regex.

After this pull request, I should be able to get into the different places where RStan does that and remove make it not rely on regex.